clang/include/clang/CIR/Dialect/IR/CIROps.td

```diff
@@ -547,6 +547,52 @@ def ScopeOp : CIR_Op<"scope", [DeclareOpInterfaceMethods<RegionBranchOpInterface
   ];
 }
 
+//===----------------------------------------------------------------------===//
+// UnaryOp
+//===----------------------------------------------------------------------===//
+
+def UnaryOpKind_Inc : I32EnumAttrCase<"Inc", 1, "inc">;
+def UnaryOpKind_Dec : I32EnumAttrCase<"Dec", 2, "dec">;
+
+def UnaryOpKind : I32EnumAttr<
+    "UnaryOpKind",
+    "unary operation kind",
+    [UnaryOpKind_Inc,
+     UnaryOpKind_Dec]> {
+  let cppNamespace = "::mlir::cir";
+}
+
+// FIXME: NoSideEffect won't work when we add overloading.
+def UnaryOp : CIR_Op<"unary",
+                     [NoSideEffect,
+                      SameOperandsAndResultType]> {
+  let summary = "Unary operations";
+  let description = [{
+    `cir.unary` performs the unary operation according to
+    the specified opcode kind: [inc, dec].
+
+    Note for inc and dec correspond only to the addition/subtraction,
+    not to the memory operations of the prefix/postfix inc/dec.
+
+    It requires one input operand and has one result, both types
+    should be the same.
+
+    ```mlir
+    %7 = cir.mut(inc, %1) : i32 -> i32
+    %8 = cir.mut(dec, %2) : i8 -> i8
+    ```
+  }];
+
+  let results = (outs AnyType:$result);
+  let arguments = (ins Arg<UnaryOpKind, "unary op kind">:$kind, Arg<AnyType>:$input);
+
+  let assemblyFormat = [{
+      `(` $kind `,` $input `)` `:` type($input) `,` type($result) attr-dict
+  }];
+
+  // TODO: Add verifier.
```

Comment:
Thanks for working on this, awesome!

This is a good way to start the mapping of such unarys, we can always revisit these in case we find interesting usecases that would require us to have a single operation with implicit loads/store semantics.

Can you please add a verifier for inc/dec as part of this patch? We can verify that (1) source value always comes from a cir.load and (2) resulting value is a stored. This is also a way to define/hardcode the intended semantics. If needed, a prepare transformation can transform that to cir.binop(add, 1) or cir.binop(sub, 1), which will make it 

Also state in the docs that for inc/dec we restrict that the value must be immediately loaded/stored.

Reply:
Changed to the load - inc/dec - store combo. Verifier added to check for this.